### ISSUE TITLE
WindowData option to consider 

### ISSUE THREAD

**Issue by pavlis** [mspass-team/mspass#551]
In working on a new multichannel cross-correlation gizmo I realized a minor issue with our `WindowData` function.  It is currently completely dogmatic in handling a situation that probably should not be viewed as black and white.   The issue is what should `WindowData` do when the interval requested is bigger than the range of the data being windowed?  That could be either a starttime or endtime inconsistency.  Currently in either case `WindowData` will kill the output and log a message to elog.  

I would suggest we add an option to `WindowData` to be less dogmatic in this situation.   The current function signature is this:
```
def WindowData(d, win_start, win_end, t0shift=None, 
   object_history=False, alg_name='scale', alg_id=None, dryrun=False)
```
I suggest it become this:
```
def WindowData(d, win_start, win_end, t0shift=None, 
   # suggested additions 
    short_segment_handling=["kill","log"],
    object_history=False, alg_name='scale', alg_id=None, dryrun=False)
```

The name `short_segment_handling` as suggest may be a bit to complicated.  The idea is the pair of keywords define how this problem is handled (the alternative is two kwarg with different names).   The idea is that component 0 defines how the issue should be handled and the second the algorithm for recovery if appropriate.  The acceptable options would be

1. ["kill","log"] - the current behavior and the clear default to be backward compatible.
2. ["truncate","log"] truncate the output if required but log a complaint
3. ["truncate","silent"] truncate silently
4. ["pad","log"] zero pad undefined section to return the right number of samples expected from window range and log as complaint.
5. ["pad","silent"] same as 4 but do it silently
6. ["gap","log"] - return result as a `TimeSeriesWGap` object with nonoverlap defined as a gap.  Log a complaint
7. ["gap","silent"] - like 6 but do it silently.

That sounds complicated but would be pretty easy to implement.   It probably should be done as two different arguments rather than the pair as suggested.  0 is basically how to handle the problem and 1 is whether or not it should initiate an error log message.  Hence, rather than a pair when I think about this it should probably be:
```
def WindowData(d, win_start, win_end, t0shift=None, 
   # suggested additions 
    short_segment_handling="kill",
    log_recoverable_errors=True,
    object_history=False, alg_name='scale', alg_id=None, dryrun=False)
```
Noting name implies proper behavior here that only recoverable errors should be allowed to be handled silently.   

I can do this change as part of the extensive changes I'm doing anyway for the multichannel correlation and new deconvolution algorithm with very little effort if the group thinks this is a good idea.

**Comment by pavlis** [comment]
Have been looking at this some more, and have a related suggestion.  Our current code base has this function defined:
```
def WindowData_with_duration(
    d,
    duration,
    t0shift=None,
    object_history=False,
    alg_name="scale",
    alg_id=None,
    dryrun=False,
):
```
This function is completely unnecessary as exactly the same functionality would be possible with:
```
d = WindowData(d,d.t0,d.t0+duration)
```
In fact, the current code has an error.  Near the top is this code fragment:
```
    win_start = d.t0 + 1
    win_end = win_start + duration
    if d.dead():
        return d
    twcut = TimeWindow(win_start, win_end)
```
win_start is WRONG as it arbitrarily adds 1 s to the data start time to define the window.   Whoever wrote this function got data index numbers and time mixed up in the api.   Actually, even that would be wrong as an arbitrary offset of 1 sample would as weird as 1 s but just less noticable.

I suggest we  deprecate this function or remove it entirely from the repository.   I am certain it only appears in our test suite and it would be trivial to change all occurrences to the form I noted above.   There is little reason for it to contaminate our code base.

**Comment by wangyinz** [comment]
well... I can see that that code has been there for 2 years. I don't understand what the +1 here is trying to do neither. Anyway, as you suggested, we can remove the function.

**Comment by wangyinz** [comment]
I just looked into it a bit more and can see that the function was added by #300. I wonder if the aws workflow was using this all the time, but then, I don't see that in the aws notebooks.

**Comment by pavlis** [comment]
I have created an as yet untested prototype for the fixes suggested here on a branch I just pushed called `extend_window_data`.   We should discuss my suggested changes in our weekly meeting today.  If the group approves I'll then proceed to modify the tests to address the changes.  For the record the changes I'm suggesting are:
1.   Add the functionality noted above to handle data time segments where the data time range is smaller than the requested time window range passed to the `WindowData` function.   I have a prototype implemented for both the "truncate" and "pad" options.   I used the second api suggestion with a list of method for handling and boolean controlling if the corrections are made silently or generate an error log message. 
2. I extended the WindowData function to work for ensembles with the same options.  In the process I created an intermediary I called `WindowDataAtomic` to, as the name suggests, handle atomic data.   The `WindowData` function then handles atomic data by a simple direct pass through and ensembles are handled by a loop calling the `WindowDataAtomic` function on each member. 
3. I removed `WindowData_with_duration`.   

I emphasize the pytest function for this module will now most definitely break so I've not issued a pull request yet.  Item 3, in particular, needs to be decided before I can proceed to fix the test functions,.
